File: CSES_PROBLEM/CoinWays.cpp
#include<bits/stdc++.h>
using namespace std;

int main(){
    
    long long int mod = 1000000007;
    long long int n,sum;
    cin>>n>>sum;
    vector<long long> coin(n);
    for (int i = 0; i < n; i++)
        cin>>coin[i];
    vector<long long int> dp(sum+1,0);
    dp[0] = 1;
    for (int i = 1; i < sum+1; i++)
    {
       for (int j = 0; j < n; j++)
       {
           if(i >= coin[j])
                dp[i] = (dp[i] + dp[i-coin[j]])%mod;
       }
       
    }
    cout<<dp[sum]%mod<<endl;
    
    
}

File: Revision-Contest/CoinWays.java
import java.util.Scanner;

/**Let us say that you are given a number N, you've to find the number of different ways to write it as the sum of 1, 3 and 4.

For example, if N = 5, the answer would be 6.

1 + 1 + 1 + 1 + 1
1 + 4
4 + 1
1 + 1 + 3
1 + 3 + 1
3 + 1 + 1
 */
public class CoinWays {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        long mod = 1000000007;
        int[] coin = new int[]{1,3,4};
        int sum = sc.nextInt();
        long[] dp = new long[sum+1];
        dp[0] = 1;
        for (int i = 1; i < dp.length; i++) {
            for (int j = 0; j < 3; j++) {
                if(coin[j]<=i)
                    dp[i] = (dp[i] + dp[i-coin[j]])%mod;
            }
        }
        System.out.println(dp[sum]);
        sc.close();
    }
}
